infrastructure/smart-contracts/contracts/upgrades/esXai/esXai4.sol

```diff
@@ -0,0 +1,430 @@
+        return _whitelist.at(index);
+    }
+
+    /**
+     * @dev Function to get the count of whitelisted addresses.
+     * @return The count of whitelisted addresses.
+     */
+    function getWhitelistCount() public view returns (uint256) {
+        return _whitelist.length();
+    }
+
+    /**
+     * @dev Override the transfer function to only allow addresses that are in the white list in the to or from field to go through
+     * @param to The address to transfer to.
+     * @param amount The amount to transfer.
+     */
+    function transfer(address to, uint256 amount) public override returns (bool) {
+        require(_whitelist.contains(msg.sender) || _whitelist.contains(to), "Transfer not allowed: address not in whitelist");
+        return super.transfer(to, amount);
+    }
+
+    /**
+     * @dev Override the transferFrom function to only allow addresses that are in the white list in the to or from field to go through
+     * @param from The address to transfer from.
+     * @param to The address to transfer to.
+     * @param amount The amount to transfer.
+     */
+    function transferFrom(address from, address to, uint256 amount) public override returns (bool) {
+        require(_whitelist.contains(from) || _whitelist.contains(to), "Transfer not allowed: address not in whitelist");
+        return super.transferFrom(from, to, amount);
+    }
+
+    /**
+     * @dev Function to start the redemption process
+     * @param amount The amount of esXai to redeem.
+     * @param duration The duration of the redemption process in seconds.
+     */
+    function startRedemption(uint256 amount, uint256 duration) public {
+        require(_redemptionActive, "Redemption is currently inactive");
+        require(amount > 0, "Invalid Amount");
+        require(balanceOf(msg.sender) >= amount, "Insufficient esXai balance");
+        require(duration == 15 days || duration == 90 days || duration == 180 days, "Invalid duration");
+
+        // Connect to the pool factory contract
+        PoolFactory2 poolFactory = PoolFactory2(poolFactoryAddress);
+
+        // Confirm the tx won't fail due to too many pools
+        uint256 poolCount = poolFactory.getPoolsOfUserCount(msg.sender);
+        require(poolCount <= 200, "User has too many pools. Unstake from some pools to redeem.");
+        
+        // Check if the sender failed KYC
+        bool failedKyc = PoolFactory2(poolFactoryAddress).failedKyc(msg.sender);
+        require(!failedKyc, "KYC failed, cannot redeem");
+        
+        // No longer transferring esXai from the sender's account to this contract
+        // Stoing redemption claim as voucherIssued instead
+
+        // Confirm the user has the appropriate amount of esXai available
+        uint256 currentBalance = balanceOf(msg.sender);        
+        uint256 totalEsXaiStaked = getTotalStakedEsXaiByUser(msg.sender);
+        uint256 totalEsXaiPendingRedemption = getTotalEsXaiPendingRedemptions(msg.sender);
+
```

Comment:
I understand these two util functions could have been omitted in favor of executing the logic here, but I thought.

1. Those functions might be valuable to have to consume that data on the front end.
2. It's substantially more readable this way.

If you believe the tradeoff is not worth separating them, I can re-factor.